operator/pkg/controllers/etcd/etcd.go

```diff
@@ -20,181 +20,77 @@ import (
-	statefulSet := &appsv1.StatefulSet{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      etcdServiceNameFor(controlPlane.ClusterName()),
-			Namespace: controlPlane.NamespaceName(),
-			OwnerReferences: []metav1.OwnerReference{{
-				APIVersion: controlPlane.APIVersion,
-				Name:       controlPlane.Name,
-				Kind:       controlPlane.Kind,
-				UID:        controlPlane.UID,
-			}},
+	rootCA := rootCACertConfig(object.NamespacedName(caSecretNameFor(cp.ClusterName()), cp.NamespaceName()))
+	secretTreeMap := common.CertTree{
+		rootCA: {
+			etcdServerCertConfig(cp),
+			etcdPeerCertConfig(cp),
 		},
 	}
-	result, err := controllerutil.CreateOrPatch(ctx, c.kubeClient, statefulSet, func() (err error) {
-		// Generate the default pod spec for the given control plane, if user has
-		// provided custom config for the etcd pod spec, patch this user
-		// provided config to the default spec
-		etcdSpec, err := patch.PodSpec(PodSpecFor(controlPlane), controlPlane.Spec.Etcd.Spec)
-		if err != nil {
-			return fmt.Errorf("failed to patch pod spec, %w", err)
-		}
-		statefulSet.Spec = appsv1.StatefulSetSpec{
-			Selector: &metav1.LabelSelector{
-				MatchLabels: etcdLabelFor(controlPlane.ClusterName()),
-			},
-			ServiceName: etcdServiceNameFor(controlPlane.ClusterName()),
-			Replicas:    aws.Int32(defaultEtcdReplicas),
-			Template: v1.PodTemplateSpec{
-				ObjectMeta: metav1.ObjectMeta{
-					Labels: etcdLabelFor(controlPlane.ClusterName()),
-				},
-				Spec: etcdSpec,
-			},
-		}
-		return nil
-	})
-	if err != nil {
-		return err
-	}
-	zap.S().Infof("[%s] statefulset %s %s", controlPlane.ClusterName(), statefulSet.Name, result)
-	return nil
-}
-
-func etcdServerPortNameFor(clusterName string) string {
-	return fmt.Sprintf("etcd-server-ssl-%s", clusterName)
-}
-
-func etcdClientPortNameFor(clusterName string) string {
-	return fmt.Sprintf("etcd-client-ssl-%s", clusterName)
-}
-
-func etcdServiceNameFor(clusterName string) string {
-	return fmt.Sprintf("%s-etcd", clusterName)
+	return c.certificates.ReconcileFor(ctx, secretTreeMap, cp)
 }
 
-func etcdCASecretNameFor(clusterName string) string {
+func caSecretNameFor(clusterName string) string {
 	return fmt.Sprintf("%s-etcd-ca", clusterName)
 }
 
-func etcdServerSecretNameFor(clusterName string) string {
+func serverSecretNameFor(clusterName string) string {
 	return fmt.Sprintf("%s-etcd-server", clusterName)
 }
 
 func etcdPeerSecretNameFor(clusterName string) string {
 	return fmt.Sprintf("%s-etcd-peer", clusterName)
 }
 
-func etcdLabelFor(clusterName string) map[string]string {
-	return map[string]string{
-		"app": etcdServiceNameFor(clusterName),
-	}
-}
-
-func certListFor(controlPlane *v1alpha1.ControlPlane) []*secrets.Request {
-	return []*secrets.Request{
-		etcdServerCertConfig(controlPlane),
-		etcdPeerCertConfig(controlPlane),
+func rootCACertConfig(nn types.NamespacedName) *secrets.Request {
+	return &secrets.Request{
```

Comment:
If you wanted to be really fancy, secrets.Request could take a namespacedname instead of separating the fields, so you could do

```
secrets.Request { NamespacedName: nn, ... } 
```